cmd/traffic_manager/metrics.cc

```diff
@@ -0,0 +1,385 @@
+
+  // If not, get the record data type.
+  if (RecGetRecordDataType(rec_name, &data_type) == REC_ERR_FAIL) {
+    luaL_error(L, "unknown metric name '%s'", rec_name);
+  }
+
+  // Sum the cluster value.
+  if (!overviewGenerator->varClusterDataFromName(data_type, rec_name, &rec_data)) {
+    RecDataClear(data_type, &rec_data);
+
+    // If we can't get any cluster data, return nil. This will generally cause the
+    // evaluator to fail, which is handled by logging and ignoring the failure.
+    lua_pushnil(L);
+    return 1;
+  }
+
+  switch (data_type) {
+  case RECD_INT: /* fallthru */
+  case RECD_COUNTER:
+    lua_pushinteger(L, rec_data.rec_int);
+    break;
+  case RECD_FLOAT:
+    lua_pushnumber(L, rec_data.rec_float);
+    break;
+  case RECD_STRING:
+    lua_pushlstring(L, rec_data.rec_string, strlen(rec_data.rec_string));
+    break;
+  default:
+    lua_pushnil(L);
+  }
+
+  // Return 1 value on the stack.
+  return 1;
+}
+
+bool
+metrics_binding_initialize(BindingInstance &binding)
+{
+  ats_scoped_str sysconfdir(RecConfigReadConfigDir());
+  ats_scoped_str config(Layout::get()->relative_to(sysconfdir, "metrics.config"));
+
+  if (!binding.construct()) {
+    mgmt_fatal(stderr, 0, "failed to initialize Lua runtime\n");
+  }
+
+  // Register the metrics userdata type.
+  lua_metrics_register(binding.lua);
+  update_metrics_namespace(binding.lua);
+
+  // Register our own API.
+  binding.bind_function("integer", metrics_create_integer);
+  binding.bind_function("counter", metrics_create_counter);
+  binding.bind_function("float", metrics_create_float);
+  binding.bind_function("metrics.cluster.sum", metrics_cluster_sum);
+
+  binding.bind_constant("metrics.now.msec", timestamp_now_msec());
+  binding.bind_constant("metrics.update.pass", int64_t(0));
+
+  // Stash a backpointer to the evaluators.
+  binding.attach_ptr("evaluators", new EvaluatorList());
+
+  // Finally, execute the config file.
+  if (binding.require(config.get())) {
+    return true;
+  }
+
+  return false;
+}
+
+void
+metrics_binding_destroy(BindingInstance &binding)
+{
+  EvaluatorList *evaluators;
+
+  evaluators = (EvaluatorList *)binding.retrieve_ptr("evaluators");
+  binding.attach_ptr("evaluators", NULL);
+  delete evaluators;
+}
+
+void
+metrics_binding_evaluate(BindingInstance &binding)
+{
+  EvaluatorList *evaluators;
+
+  evaluators = (EvaluatorList *)binding.retrieve_ptr("evaluators");
+  ink_release_assert(evaluators != NULL);
+
+  // Keep updating the namespace until it settles (ie. we make 0 updates).
+  if (evaluators->update) {
+    evaluators->update = update_metrics_namespace(binding.lua) ? true : false;
+  }
+
+  binding.bind_constant("metrics.now.msec", timestamp_now_msec());
+  binding.bind_constant("metrics.update.pass", ++evaluators->passes);
+  evaluators->evaluate(binding.lua);
+
+  // Periodically refresh the namespace to catch newly added metrics.
+  if (evaluators->passes % 10 == 0) {
```

Comment:
This is dealing with the problem that metrics are not all available at startup. They are created over time as subsystems initialize, and can be created at any time by plugins. So while metrics are being created we keep trying to install them in the Lua namespace, but once we finish we just periodically double check. The 10 iterations is completely arbitrary.